src/utils/view.js

```diff
@@ -0,0 +1,335 @@
+import 'd3-transition';
+import { event, select } from 'd3-selection';
+import { interpolate } from 'd3-interpolate';
+import { zoom as d3Zoom, zoomIdentity, zoomTransform } from 'd3-zoom';
+
+/**
+ * Applies an active viewport behaviour on the given elements using d3-zoom.
+ * Returns an object representing the view for use in view related functions.
+ * @param {Object} options The options to use
+ * @param {React.Ref} options.container The container element ref
+ * @param {React.Ref} options.wrapper The wrapper element ref
+ * @param {?Function} options.onViewChanged Callback when view changes
+ * @param {?Function} options.onViewEnd Callback when view change ends
+ * @param {?Boolean} [options.allowUserInput=true] Enable or disable user input
+ * @returns {Object} An object representing the view
+ */
+export const viewing = ({
+  container,
+  wrapper,
+  onViewChanged,
+  onViewEnd,
+  allowUserInput = true
+}) => {
+  const zoom = d3Zoom()
+    .on('zoom', () => {
+      const transform = event.transform;
+
+      // Ignore invalid transforms
+      if (isInvalidTransform(transform)) {
+        return;
+      }
+
+      select(wrapper.current).attr('transform', transform);
+      onViewChanged && onViewChanged(transform);
+    })
+    .on('end', onViewEnd)
+    // Use linear interpolation
+    .interpolate(interpolate);
+
+  if (!allowUserInput) {
+    // Ignore all user input default behaviour
+    zoom.filter(() => false);
+  }
+
+  select(container.current)
+    .call(zoom)
+    // Disable default double click to avoid conflicts
+    .on('dblclick.zoom', null);
+
+  return {
+    zoom,
+    container,
+    wrapper
+  };
+};
+
+/**
+ * A constant for the origin transform { x: 0, y: 0, k: 1 }
+ */
+export const origin = zoomIdentity;
+
+/**
+ * Returns true if the transform is equivalent to the origin { x: 0, y: 0, k: 1 }
+ * @param {Object} transform A transform object to test
+ * @returns {Boolean} True if equivalent else false
+ */
+export const isOrigin = ({ x, y, k }) =>
+  x === origin.x && y === origin.y && k === origin.k;
+
+/**
+ * Returns true if the transform meets this definition:
+ *  - any component (x, y or k) is infinite, `NaN` or `undefined`
+ * otherwise returns false.
+ * @param {Object} transform A transform object to test
+ * @returns {Boolean} True if the transform is invalid else false
+ */
+export const isInvalidTransform = ({ x, y, k }) =>
+  !isFinite(x + y + k) || isNaN(x + y + k);
+
+/**
+ * Returns the current transform of the given view.
+ * This is equivalent to the container CSS transform but negated.
+ * @param {Object} view The view
+ * @returns {Object} The view transform
+ */
+export const getViewTransform = view => {
+  const transform = zoomTransform(view.wrapper.current);
+  return isInvalidTransform(transform) ? origin : negateTransform(transform);
+};
+
+/**
+ * Negates the x and y components, in effect making the direction opposite.
+ * Scale component does not change. Zeroes are always positive.
```

Comment:
Are we ending comments with full stops within the code base as I noticed that it has not been applied elsewhere?

Reply:
I don't think it's something we're picky about. It's not public-facing so I don't think it matters, as long as its meaning is clear to readers

Reply:
Agreed, I usually tend to include them in JSDocs written over multiple sentences, but not usually in comments intended as short notes rather than full sentences, so best to just judge on clarity and brevity.

Reply:
Exactly, same.